xds/reloading_stream.go

```diff
@@ -0,0 +1,108 @@
+package xds
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	v3discoverypb "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/grpc"
+
+	"github.com/gojek/courier-go/xds/backoff"
+	"github.com/gojek/courier-go/xds/log"
+)
+
+type adsStream v3discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient
+
+type reloadingStream struct {
+	mu       sync.RWMutex
+	reloadWG sync.WaitGroup
+	log      log.Logger
+	s        adsStream
+	ns       func() v3discoverypb.AggregatedDiscoveryServiceClient
+
+	strategy    backoff.Strategy
+	onReConnect func(error)
+	reloadCh    chan error
+	connTimeout time.Duration
+}
+
+func (r *reloadingStream) Send(req *v3discoverypb.DiscoveryRequest) error {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.s.Send(req)
+}
+
+func (r *reloadingStream) Recv() (*v3discoverypb.DiscoveryResponse, error) {
+	resp, err := r.safeRead()
+	if err == nil {
+		return resp, nil
+	}
+
+	r.reloadWG.Add(1)
+	r.reloadCh <- err
+	r.reloadWG.Wait()
+
+	return r.safeRead()
+}
+
+func (r *reloadingStream) safeRead() (*v3discoverypb.DiscoveryResponse, error) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.s.Recv()
+}
+
+func (r *reloadingStream) createStream() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), r.connTimeout)
+	defer cancel()
+
+	stream, err := r.ns().StreamAggregatedResources(ctx, grpc.WaitForReady(true))
+	if err != nil {
+		return err
+	}
+
+	r.s = stream
+
+	return nil
+}
+
+func (r *reloadingStream) startReloader(ctx context.Context) {
+	retries := 0
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case err := <-r.reloadCh:
+			retries++
+
+			if r.onReConnect != nil {
```

Comment:
onReconnect should be called after new stream has been successfully created.

Reply:
I want to call the function before reinitializing the stream since that was the previous behavior where `version` and `nonce` were reset. I will rename this function to indicate the execution point better.

Reply:
Actually, you are correct, if we call this after the stream creation we can avoid multiple calls to onReconnect. I will move this call.